File: sdk/python/example/example_api.py
import logging
import os
import uuid
from datetime import datetime, timedelta

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.generate.account.fee.model_get_spot_actual_fee_req import GetSpotActualFeeReqBuilder
from kucoin_universal_sdk.generate.futures.market.model_get_klines_req import GetKlinesReqBuilder, GetKlinesReq
from kucoin_universal_sdk.generate.service.account_api import AccountService
from kucoin_universal_sdk.generate.service.futures_api import FuturesService
from kucoin_universal_sdk.generate.service.spot_api import SpotService
from kucoin_universal_sdk.generate.spot.order.model_add_order_sync_req import AddOrderSyncReqBuilder, AddOrderSyncReq
from kucoin_universal_sdk.generate.spot.order.model_cancel_order_by_order_id_sync_req import \
    CancelOrderByOrderIdSyncReqBuilder
from kucoin_universal_sdk.generate.spot.order.model_get_order_by_order_id_req import GetOrderByOrderIdReqBuilder
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model.transport_option import TransportOptionBuilder


def rest_example():
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )

    #  Retrieve API secret information from environment variables
    key = os.getenv("API_KEY")
    secret = os.getenv("API_SECRET")
    passphrase = os.getenv("API_PASSPHRASE")

    # Optional: Retrieve broker secret information from environment variables; applicable for broker API only
    broker_name = os.getenv("BROKER_NAME")
    broker_key = os.getenv("BROKER_KEY")
    broker_partner = os.getenv("BROKER_PARTNER")

    # Set specific options, others will fall back to default values
    http_transport_option = (
        TransportOptionBuilder()
        .set_keep_alive(True)
        .set_max_pool_size(10)
        .set_max_connection_per_pool(10)
        .build()
    )

    # Create a client using the specified options
    client_option = (
        ClientOptionBuilder()
        .set_key(key)
        .set_secret(secret)
        .set_passphrase(passphrase)
        .set_broker_name(broker_name)
        .set_broker_key(broker_key)
        .set_broker_partner(broker_partner)
        .set_spot_endpoint(GLOBAL_API_ENDPOINT)
        .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
        .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
        .set_transport_option(http_transport_option)
        .build()
    )
    client = DefaultClient(client_option)

    # Get the Restful Service
    kucoin_rest_service = client.rest_service()

    account_service_example(kucoin_rest_service.get_account_service())
    spot_service_example(kucoin_rest_service.get_spot_service())
    futures_service_example(kucoin_rest_service.get_futures_service())


def account_service_example(account_service: AccountService):
    # Get account api
    account_api = account_service.get_account_api()
    # Get summary information
    account_info_resp = account_api.get_account_info()
    logging.info(f"account info: level:{account_info_resp.level}, SubAccountSize:{account_info_resp.sub_quantity}")
    # Other Account APIs...

    # Get fee api
    fee_api = account_service.get_fee_api()
    get_actual_fee_req = GetSpotActualFeeReqBuilder().set_symbols("BTC-USDT,ETH-USDT").build()
    get_actual_fee_resp = fee_api.get_spot_actual_fee(get_actual_fee_req)
    for fee_data in get_actual_fee_resp.data:
        logging.info(
            f"fee info: symbol:{fee_data.symbol}, TakerFee:{fee_data.taker_fee_rate}, MakerFee:{fee_data.maker_fee_rate}")
    # Other Fee APIs...

    #  Other APIs related to account...


def spot_service_example(spot_service: SpotService):
    order_api = spot_service.get_order_api()

    # Add limit order
    add_order_req = (AddOrderSyncReqBuilder()
                     .set_client_oid(str(uuid.uuid4()))
                     .set_side(AddOrderSyncReq.SideEnum.BUY)
                     .set_symbol("BTC-USDT")
                     .set_type(AddOrderSyncReq.TypeEnum.LIMIT)
                     .set_remark("sdk_example")
                     .set_price("10000")
                     .set_size("0.001")
                     .build())

    resp = order_api.add_order_sync(add_order_req)
    logging.info(f"Add order success, id: {resp.order_id}, oid: {resp.order_id}")

    # Query order detail
    query_order_detail_req = (GetOrderByOrderIdReqBuilder()
                              .set_order_id(resp.order_id)
                              .set_symbol("BTC-USDT")
                              .build())
    order_detail_resp = order_api.get_order_by_order_id(query_order_detail_req)
    logging.info(f"Order detail: {order_detail_resp.to_json()}")

    # Cancel order
    cancel_order_req = (CancelOrderByOrderIdSyncReqBuilder()
                        .set_order_id(resp.order_id)
                        .set_symbol("BTC-USDT")
                        .build())
    cancel_order_resp = order_api.cancel_order_by_order_id_sync(cancel_order_req)
    logging.info(f"Cancel order success, id: {cancel_order_resp.order_id}")


def futures_service_example(futures_service: FuturesService):
    market_api = futures_service.get_market_api()

    # Query all symbols
    all_symbol_resp = market_api.get_all_symbols()
    max_query = min(len(all_symbol_resp.data), 10)

    for i in range(max_query):
        symbol = all_symbol_resp.data[i]
        # Query the Kline of the symbol
        start = int((datetime.now() - timedelta(minutes=10)).timestamp() * 1000)
        end = int(datetime.now().timestamp() * 1000)
        get_kline_req = (GetKlinesReqBuilder()
                         .set_symbol(symbol.symbol)
                         .set_granularity(GetKlinesReq.GranularityEnum.T_1)
                         .set_from_(start)
                         .set_to(end)
                         .build())
        get_kline_resp = market_api.get_klines(get_kline_req)
        rows = []
        for row in get_kline_resp.data:
            timestamp = datetime.fromtimestamp(row[0] / 1000).strftime("%Y-%m-%d %H:%M:%S")
            formatted_row = (
                f"[Time: {timestamp}, Open: {row[1]:.2f}, High: {row[2]:.2f}, "
                f"Low: {row[3]:.2f}, Close: {row[4]:.2f}, Volume: {row[5]:.2f}]"
            )
            rows.append(formatted_row)
        logging.info(f"Symbol: {symbol.symbol}, Kline: {', '.join(rows)}")


if __name__ == "__main__":
    rest_example()

File: sdk/golang/example/example_api.go
package main

import (
	"context"
	"fmt"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/account/fee"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/futures/market"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/service"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/order"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"github.com/google/uuid"
	"os"
	"strings"
	"time"
)

func RestExample() {
	// Use the default logger or supply your custom logger
	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	// Retrieve API secret information from environment variables
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	// Optional: Retrieve broker secret information from environment variables; applicable for broker API only
	brokerName := os.Getenv("BROKER_NAME")
	brokerKey := os.Getenv("BROKER_KEY")
	brokerPartner := os.Getenv("BROKER_PARTNER")

	// Set specific options, others will fall back to default values
	httpOption := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		SetMaxIdleConnsPerHost(10).
		Build()

	// Create a client using the specified options
	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithBrokerKey(brokerKey).
		WithBrokerName(brokerName).
		WithBrokerPartner(brokerPartner).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOption).
		Build()
	client := api.NewClient(option)

	// Get the Restful Service
	kuCoinRestService := client.RestService()

	accountServiceExample(kuCoinRestService.GetAccountService())
	spotServiceExample(kuCoinRestService.GetSpotService())
	futuresServiceExample(kuCoinRestService.GetFuturesService())

}

func accountServiceExample(accountService service.AccountService) {
	{
		// Get account api
		accountApi := accountService.GetAccountAPI()
		// Override the global timeout
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		// Get summary information
		resp, err := accountApi.GetAccountInfo(ctx)
		if err != nil {
			panic(err)
		}
		logger.GetLogger().Infof("account info: level:%d, SubAccountSize:%d", resp.Level, resp.SubQuantity)
		// Other Account APIs...
	}
	{
		// Get fee api
		feeApi := accountService.GetFeeAPI()
		// Get the actual fee rate of the trading pair
		resp, err := feeApi.GetSpotActualFee(fee.NewGetSpotActualFeeReqBuilder().SetSymbols("BTC-USDT,ETH-USDT").Build(), context.Background())
		if err != nil {
			panic(err)
		}
		for _, feeData := range resp.Data {
			logger.GetLogger().Infof("fee info: symbol:%s, TakerFee:%s, MakerFee:%s", feeData.Symbol, feeData.TakerFeeRate, feeData.MakerFeeRate)
		}
		// Other Fee APIs...
	}
	// Other APIs related to account...
}

func spotServiceExample(spotService service.SpotService) {
	orderApi := spotService.GetOrderAPI()

	// Add limit order
	addOrderReq := order.NewAddOrderSyncReqBuilder().
		SetClientOid(uuid.NewString()).
		SetSide("buy").
		SetSymbol("BTC-USDT").
		SetType("limit").
		SetRemark("sdk_example").
		SetPrice("10000").
		SetSize("0.001").
		Build()

	resp, err := orderApi.AddOrderSync(addOrderReq, context.Background())
	if err != nil {
		panic(err)
	}
	logger.GetLogger().Infof("add order success, id:%s, oid:%s", resp.OrderId, resp.OrderId)

	// Query order detail
	queryOrderDetailReq := order.NewGetOrderByOrderIdReqBuilder().SetOrderId(resp.OrderId).SetSymbol("BTC-USDT").Build()
	orderDetailResp, err := orderApi.GetOrderByOrderId(queryOrderDetailReq, context.Background())
	if err != nil {
		panic(err)
	}
	logger.GetLogger().Infof("order detail: %+v", orderDetailResp.ToMap())

	// Cancel order
	cancelOrderReq := order.NewCancelOrderByOrderIdSyncReqBuilder().SetOrderId(resp.OrderId).SetSymbol("BTC-USDT").Build()
	cancelOrderResp, err := orderApi.CancelOrderByOrderIdSync(cancelOrderReq, context.Background())
	if err != nil {
		panic(err)
	}
	logger.GetLogger().Infof("cancel order success, id:%s", cancelOrderResp.OrderId)
}

func futuresServiceExample(futuresService service.FuturesService) {
	marketApi := futuresService.GetMarketAPI()

	// Query all symbols
	allSymbolResp, err := marketApi.GetAllSymbols(context.Background())
	if err != nil {
		panic(err)
	}

	maxQuery := min(len(allSymbolResp.Data), 10)
	for i := 0; i < maxQuery; i++ {
		symbol := allSymbolResp.Data[i]
		// Query the Kline of the symbol
		start := time.Now().Add(-time.Minute*10).UnixNano() / int64(time.Millisecond)
		end := time.Now().UnixNano() / int64(time.Millisecond)
		getKlineReq := market.NewGetKlinesReqBuilder().SetSymbol(symbol.Symbol).SetGranularity(1).SetFrom(start).SetTo(end).Build()
		getKlineResp, err := marketApi.GetKlines(getKlineReq, context.Background())
		if err != nil {
			panic(err)
		}
		var rows []string
		for _, row := range getKlineResp.Data {
			timestamp := time.Unix(int64(row[0]/1000), 0).Format("2006-01-02 15:04:05")

			formattedRow := fmt.Sprintf(
				"[Time: %s, Open: %.2f, High: %.2f, Low: %.2f, Close: %.2f, Volume:%.2f]",
				timestamp, row[1], row[2], row[3], row[4], row[5],
			)
			rows = append(rows, formattedRow)
		}
		logger.GetLogger().Infof("symbol:%s, kline:%+v", symbol.Symbol, strings.Join(rows, ","))
	}
}
